Handle unbacked SymInt sized outputs in AOTAutograd

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113159

Thanks aakhundov for constructing the test case. This PR was constructed by running the failing test case, and then fixing problems until we got all the way to the end. There are a few distinct fixes:

* AOTAutograd performs equality tests on tensor metadata to determine if a metadata mutation had occurred. If we test i0 vs i1, we should report these are NOT equal, since obviously we have somehow resized the tensor from i0 to i1 (even if, on a particular run, it is possible i0 == i1).
* There's a sketchy fix for `test_aot_autograd_exhaustive_matmul_cpu_float32` where we check if the output shape equals the tangent shape. Unfortunately, the same `definitely_true` treatment does not work here, it still fails on the example. I piled an extra sketchy fix on top of it, where I just try my best to avoid doing the view. Maybe we should have some sort of logging here.
* Partitioner needs to get out a size for unbacked SymInt when partitioning. I just feed it a random heuristic value in this case, similar to how we've been dealing with this in Inductor.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang committed:
Handle unbacked SymInt sized outputs in AOTAutograd

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

aakhundov commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -274,6 +279,20 @@ def parallel_and(*args):
+    Like ==, but when run on list/tuple, it will recursively test equality
+    and use sym_and to join the results together, without guarding.
+    """
+    if (isinstance(x, tuple) and isinstance(y, tuple)) or (isinstance(x, list) and isinstance(y, list)):
```
Can it be that one is list and other is tuple; and do we want to allow that, too?

ezyang replied:
I'm matching `==` semantics, which says that list and tuple are never equal

aakhundov reviewed (approved):
Thanks for the quick fix!

ezyang committed:
Update on "Handle unbacked SymInt sized outputs in AOTAutograd"


Thanks aakhundov for constructing the test case. This PR was constructed by running the failing test case, and then fixing problems until we got all the way to the end. There are a few distinct fixes:

* AOTAutograd performs equality tests on tensor metadata to determine if a metadata mutation had occurred. If we test i0 vs i1, we should report these are NOT equal, since obviously we have somehow resized the tensor from i0 to i1 (even if, on a particular run, it is possible i0 == i1).
* There's a sketchy fix for `test_aot_autograd_exhaustive_matmul_cpu_float32` where we check if the output shape equals the tangent shape. Unfortunately, the same `definitely_true` treatment does not work here, it still fails on the example. I piled an extra sketchy fix on top of it, where I just try my best to avoid doing the view. Maybe we should have some sort of logging here.
* Partitioner needs to get out a size for unbacked SymInt when partitioning. I just feed it a random heuristic value in this case, similar to how we've been dealing with this in Inductor.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang committed:
Update on "Handle unbacked SymInt sized outputs in AOTAutograd"


Thanks aakhundov for constructing the test case. This PR was constructed by running the failing test case, and then fixing problems until we got all the way to the end. There are a few distinct fixes:

* AOTAutograd performs equality tests on tensor metadata to determine if a metadata mutation had occurred. If we test i0 vs i1, we should report these are NOT equal, since obviously we have somehow resized the tensor from i0 to i1 (even if, on a particular run, it is possible i0 == i1).
* There's a sketchy fix for `test_aot_autograd_exhaustive_matmul_cpu_float32` where we check if the output shape equals the tangent shape. Unfortunately, the same `definitely_true` treatment does not work here, it still fails on the example. I piled an extra sketchy fix on top of it, where I just try my best to avoid doing the view. Maybe we should have some sort of logging here.
* Partitioner needs to get out a size for unbacked SymInt when partitioning. I just feed it a random heuristic value in this case, similar to how we've been dealing with this in Inductor.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -314,10 +314,13 @@ def _size_of(node: fx.Node) -> int:
                 return 999999
+        # NB: The fallback values here are meaningless, maybe we should respect
+        # torch._inductor.config.unbacked_symint_fallback (but this is a
+        # layering violation)
```
Instead of picking a random default, should we the value ranges upper bound if one exists? (I think it's something like `hint_int(shape_env.var_to_range[val.node.expr].upper)`)?

Also, just curious since you're picking a random value here - do you think this value won't matter in practice? (that we are unlikely to run into models where some tensor with an unbacked size could optionally be saved in the bw vs recomputed). Do you think we should just introduce a new config to avoid the layering issue?
```
torch._functorch.config.unbacked_symint_fallback = 8192
torch._inductor.config.unbacked_symint_fallback = torch._functorch.config.unbacked_symint_fallback
```

ezyang replied:
> Instead of picking a random default, should we the value ranges upper bound if one exists? (I think it's something like hint_int(shape_env.var_to_range[val.node.expr].upper))?

We should. Doing this in general is somewhat involved, so we should probably do it in its own PR.

> Also, just curious since you're picking a random value here - do you think this value won't matter in practice? (that we are unlikely to run into models where some tensor with an unbacked size could optionally be saved in the bw vs recomputed).

I think it can eventually matter in complicated enough situations. But there are some countervailing factors: (1) we can also make suboptimal decisions when hints do exist--if we compile with a hint that is not representative of the true cost of something, and (2) if the user doesn't warmup on representative inputs, the correct range here is fundamentally unknowable. Even the upper bound, if it exists, is merely that: an upper bound; it may be that you expect it to be half the size of the theoretical upper bound, and that will make a big difference. So in the end, to handle this correctly, we will need some sort of user annotation or profile-guided optimization approach. So I'm not *too* worried about this as a stopgap right now.

> Do you think we should just introduce a new config to avoid the layering issue?

Maybe. Actually I think @peterbell10 added a new dedicated ShapeEnv config at https://github.com/pytorch/pytorch/pull/112933 so we can probably just centralize it all there.

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -274,6 +279,20 @@ def parallel_and(*args):
+    if (isinstance(x, tuple) and isinstance(y, tuple)) or (isinstance(x, list) and isinstance(y, list)):
+        if len(x) != len(y):
+            return False
+        return functools.reduce(operator.and_, map(sym_eq, x, y), True)
```
I kind of thought that the default implementation of `==` would do this for lists, but I guess that's wrong

ezyang replied:
The default == implementation is short circuiting: 

```
>>> class A:
...   def __eq__(self, other):
...     assert False
... 
>>> 
>>> [1, A()] == [2, A()]
False
>>> [1, A()] == [1, A()]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in __eq__
AssertionError
```

ezyang committed:
Update on "Handle unbacked SymInt sized outputs in AOTAutograd"


Thanks aakhundov for constructing the test case. This PR was constructed by running the failing test case, and then fixing problems until we got all the way to the end. There are a few distinct fixes:

* AOTAutograd performs equality tests on tensor metadata to determine if a metadata mutation had occurred. If we test i0 vs i1, we should report these are NOT equal, since obviously we have somehow resized the tensor from i0 to i1 (even if, on a particular run, it is possible i0 == i1).
* There's a sketchy fix for `test_aot_autograd_exhaustive_matmul_cpu_float32` where we check if the output shape equals the tangent shape. Unfortunately, the same `definitely_true` treatment does not work here, it still fails on the example. I piled an extra sketchy fix on top of it, where I just try my best to avoid doing the view. Maybe we should have some sort of logging here.
* Partitioner needs to get out a size for unbacked SymInt when partitioning. I just feed it a random heuristic value in this case, similar to how we've been dealing with this in Inductor.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
